### PR TITLE
Added configuration types

### PR DESCRIPTION
**Pull request by zsweigart**
Added types needed to configure and open link for both public key and link token configurations.

### CODE REVIEW

**Review by amytang0 — APPROVED** [review]
LGTM

File: types/Types.d.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    webhook?: string;
+    linkCustomizationName?: string;
+    accountSubtypes?: Array<LinkAccountSubtype>;
+    oauthNonce?: string;
+    oauthRedirectUri?: string;
+    oauthStateId?: string;
+});
+
+export enum LinkLogLevel {
+    ERROR,
```
Comment (amytang0) [inline types/Types.d.ts:33]: nit: Kind of expected least severe to most, but could just be me

File: types/Types.d.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+export enum PlaidProduct {
+    ASSETS,
+    AUTH,
+    IDENTITY,
+    INCOME,
+    INVESTMENTS,
+    LIABILITIES,
+    LIABILITIES_REPORT,
+    PAYMENT_INITIATION,
+    TRANSACTIONS
```
Comment (amytang0) [inline types/Types.d.ts:54]: nit: trailing comma is nice for blame

File: types/Types.d.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+export enum LinkAccountType {
+    CREDIT = 'credit',
+    DEPOSITORY = 'depository',
+    INVESTMENT = 'investment',
+    LOAN = 'loan',
+    OTHER = 'other'
```
Comment (amytang0) [inline types/Types.d.ts:62]: nit: trailing comma is nice for blame

**Review by amytang0 — DISMISSED** [review]
LGTM, small nits

**Review by amytang0 — DISMISSED** [review]
LGTM

File: types/Types.d.ts [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+    extraParams?: Record<string, any>;
+}
+
+export type LinkTokenConfiguration = (CommonPlaidLinkOptions & {
+    token: string;
+});
+
+export type LinkPublicKeyConfiguration = (CommonPlaidLinkOptions & {
+    publicKey: string;
+    token?: string;
```
Comment (JeroenMols) [inline types/Types.d.ts:14]: Order:should be

```

class LinkPublicKeyConfiguration private constructor(
  val token: String? = null
  val publicKey: String? = null
  val clientName: String? = null
  val environment: PlaidEnvironment = PlaidEnvironment.SANDBOX
  val products: List<PlaidProduct> = emptyList()
  val language: String = Locale.ENGLISH.language
  val countryCodes: List<String> = listOf(Locale.US.country)
  val webhook: String? = null
   val userLegalName: String? = null
   val userEmailAddress: String? = null
   val userPhoneNumber: String? = null
   val linkCustomizationName: String? = null
   val accountSubtypes: List<LinkAccountSubtype>? = null
   val logLevel: LinkLogLevel = LinkLogLevel.ASSERT
  val extraParams: Map<String String> = mapOf()
) {

```

File: types/Types.d.ts [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+    logLevel?: LinkLogLevel;
+    extraParams?: Record<string, any>;
+}
+
+export type LinkTokenConfiguration = (CommonPlaidLinkOptions & {
+    token: string;
+});
+
+export type LinkPublicKeyConfiguration = (CommonPlaidLinkOptions & {
+    publicKey: string;
```
Comment (JeroenMols) [inline types/Types.d.ts:13]:
```suggestion
    publicKey?: string;
```

Nullable on Android

File: types/Types.d.ts [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+    token?: string;
+    clientName: string;
```
Comment (JeroenMols) [inline types/Types.d.ts:15]: Nullable on Android, non null on iOS. Not sure what to do here....

File: types/Types.d.ts [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+    environment: PlaidEnvironment;
+    product: Array<PlaidProduct>;
```
Comment (JeroenMols) [inline types/Types.d.ts:17]:
```suggestion
    products: Array<PlaidProduct>;
```

lists/arrays should have plural names

File: types/Types.d.ts [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    countryCodes?: Array<string>;
```
Comment (JeroenMols) [inline types/Types.d.ts:18]: Should be non null (is non null on both Android and iOS
```suggestion
    countryCodes: Array<string>;
```

File: types/Types.d.ts [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+    language?: string;
```
Comment (JeroenMols) [inline types/Types.d.ts:19]: Should be non null, is non null on both Android and iOS

```suggestion
    language: string;
```

File: types/Types.d.ts [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+    userEmailAddress?: string;
+    userLegalName?: string;
+    webhook?: string;
+    linkCustomizationName?: string;
+    accountSubtypes?: Array<LinkAccountSubtype>;
+    oauthNonce?: string;
```
Comment (JeroenMols) [inline types/Types.d.ts:25]: should oauth fields be wrapped in a `OAuthConfiguration` object like iOS?

File: types/Types.d.ts [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+export interface LinkSuccess {
+    publicToken: string;
+    metadata: LinkSuccessMetadata;
+}
+
+export interface LinkSuccessMetadata {
+    institution?: LinkInstitution;
+    accounts: LinkAccount[];
+    linkSessionId: String;
+    metadataJson: String;
```
Comment (JeroenMols) [inline types/Types.d.ts:138]: Could cause problems as this is nullable on iOS. Also make it nullable?

File: types/Types.d.ts [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+}
+
+export interface LinkAccount {
+    id: String;
+    name?: String;
+    mask?: String;
+    type: LinkAccountType;
+    subtype: LinkAccountSubtype;
+    verificationStatus: LinkAccountVerificationStatus;
```
Comment (JeroenMols) [inline types/Types.d.ts:147]: verification status is nullable on both Android and iOS

File: types/Types.d.ts (hunk 6 above)
Comment (zsweigart) [reply types/Types.d.ts:15]: It shouldn't be nullable on Android, public key config requires this

File: types/Types.d.ts (hunk 10 above)
Comment (zsweigart) [reply types/Types.d.ts:30]: makes sense, they're iOS only

File: types/Types.d.ts [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+    ERROR,
+}
+
+export enum PlaidEnvironment {
+    PRODUCTION = 'production',
+    DEVELOPMENT = 'development',
+    SANDBOX = 'sandbox',
+}
+
+export enum PlaidProduct {
```
Comment (pladaub) [inline types/Types.d.ts:48]: nit: may as well add deposit_switch here

**Review by pladaub — APPROVED** [review]
Looks good, it would be nice if we could use this to type the handoff objects at the RN Native -> JS layer

**Review by JeroenMols — APPROVED** [review]

**Pull request merged**